### PR TITLE
docs: add project layout ref page

### PR DESCRIPTION
**Pull request by agnes512**
Closes https://github.com/strongloop/loopback-next/issues/4847. See [LB3 version](https://loopback.io/doc/en/lb3/Project-layout-reference.html) for reference.

image

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/Project-layout-reference.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      <td>An example LoopBack front page.</td>
+      <td>N/A</td>
+    </tr>
+    <tr>
+      <td>Others</td>
+      <td>Add your HTML, CSS, client JavaScript files.</td>
+      <td>&nbsp;</td>
+    </tr>
+    <tr>
+      <th colspan="3" style="text-align: center; background-color: #bfbfbf;">/common directory - shared application files</th>
```
Comment (agnes512) [inline docs/site/Project-layout-reference.md:115]: Not sure if we have anything that are shared between server side and client side.

File: docs/site/Project-layout-reference.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      <td>A sub-directory for all model files. See <a href="Model.html">Model</a>.</td>
+      <td>>&nbsp;</td>
+    </tr>
+    <tr>
+      <td><code>/repositories</code> directory</td>
+      <td>A sub-directory for all repository files. See <a href="Repository.html">Repository</a>.</td>
+      <td>>&nbsp;</td>
+    </tr>
+    <tr>
+      <td><code>application.ts</code></td>
```
Comment (agnes512) [inline docs/site/Project-layout-reference.md:76]: Different from [LB3 docs](https://loopback.io/doc/en/lb3/Project-layout-reference.html), I don't have extra pages as examples for these files. Please let me know if you think they are needed.

File: docs/site/Project-layout-reference.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      <th>Description</th>
+      <th width="180">How to access in code</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th colspan="3" style="text-align: center; background-color: #bfbfbf;">Top-level application directory</th>
+    </tr>
+    <tr>
+      <td><code>/node-modules</code> directory</td>
```
Comment (bajtos) [inline docs/site/Project-layout-reference.md:34]: I find it slightly weird that some entries start with a forward slash (`/node_modules`) while other don't (`package.json`). Let's be consistent please.

To differentiate between files (`package.json`) and directories (`node_modules`), I am proposing to _append_ a slash to directory names.

```suggestion
      <td><code>node-modules/</code> directory</td>
```

File: docs/site/Project-layout-reference.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      <td>Contains Node packages as specified as dependencies in <code>package.json</code>. Update with <code>npm install</code>.</td>
```
Comment (bajtos) [inline docs/site/Project-layout-reference.md:35]: > Update with <code>npm install</code>

`npm install` typically does not update dependencies, only install missing ones. `npm update` is the command to use.

File: docs/site/Project-layout-reference.md (hunk 1 above)
Comment (bajtos) [reply docs/site/Project-layout-reference.md:115]: We don't have `common` directory in LB4, please remove this table row.

File: docs/site/Project-layout-reference.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      <th>Description</th>
+      <th width="180">How to access in code</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th colspan="3" style="text-align: center; background-color: #bfbfbf;">Top-level application directory</th>
+    </tr>
+    <tr>
+      <td><code>node-modules/</code> directory</td>
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:38]: Let's skip `node-modules` as it's common to all Node.js applications.

File: docs/site/Project-layout-reference.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+permalink: /doc/en/lb4/Project-layout-reference.html
+---
+
+LoopBack 4 project files and directories are in the _application root
+directory_. Within this directory the standard LoopBack 4 project structure has
+these sub-directories:
+
+- `src` - Node application scripts and configuration files.
+- `public` - Client JavaScript, HTML, and CSS files.
+- `definitions` - API and product definition YAML files (**IBM API Connect
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:15]: Do we have this? Let's skip it.

File: docs/site/Project-layout-reference.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Project-layout-reference.html
+---
+
+LoopBack 4 project files and directories are in the _application root
+directory_. Within this directory the standard LoopBack 4 project structure has
+these sub-directories:
+
+- `src` - Node application scripts and configuration files.
+- `public` - Client JavaScript, HTML, and CSS files.
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:14]: Client side assets (JavaScript, HTML, and CSS) for the default home page.

File: docs/site/Project-layout-reference.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Project-layout-reference.html
+---
+
+LoopBack 4 project files and directories are in the _application root
+directory_. Within this directory the standard LoopBack 4 project structure has
+these sub-directories:
+
+- `src` - Node application scripts and configuration files.
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:13]: TypeScript source code

File: docs/site/Project-layout-reference.md (hunk 7 above)
Comment (raymondfeng) [reply docs/site/Project-layout-reference.md:14]: Maybe we should mention `dist` - which contains transpiled JavaScript code generated by `npm run build`.

Comment (agnes512) [reply docs/site/Project-layout-reference.md]: I added `dist` folder under `Top level app directory`, PTAL.

File: docs/site/Project-layout-reference.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      <th>Description</th>
+      <th width="180">How to access in code</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th colspan="3" style="text-align: center; background-color: #bfbfbf;">Top-level application directory</th>
+    </tr>
+    <tr>
+      <td><code>dist/</code> directory</td>
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:37]: Let's move it to the end of top-level directories.

File: docs/site/Project-layout-reference.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      <td>Contains transpiled JavaScript code generated by <code>npm run build</code>.</td>
+      <td>npm run build</td>
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:39]: We should mention `npm run build` does incremental build. A full build can be achieved using `npm run clean && npm run build`.

File: docs/site/Project-layout-reference.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    <tr>
+      <th colspan="3" style="text-align: center; background-color: #bfbfbf;">/src directory - Node application files</th>
+    </tr>
+    <tr>
+      <td><a href="https://github.com/strongloop/loopback-next/tree/master/examples/todo/src/controllers"><code>controllers/</code></a> directory</td>
+      <td>A sub-directory for all controller files. See <a href="Controller.html">Controller</a>.</td>
+      <td>&nbsp;</td>
+    </tr>
+    <tr>
+      <td><a href="https://github.com/strongloop/loopback-next/tree/master/examples/todo/src/datasources"><code>datasources/</code></a> directory</td>
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:69]: Let's remove `directory` from the description as the trailing `/` already tells so. This comment should apply to other rows too.

File: docs/site/Project-layout-reference.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      <td><a href="https://github.com/strongloop/loopback-next/tree/master/examples/todo/src/models"><code>models/</code></a> directory</td>
+      <td>A sub-directory for all model files. See <a href="Model.html">Model</a>.</td>
+      <td>&nbsp;</td>
+    </tr>
+    <tr>
+      <td><a href="https://github.com/strongloop/loopback-next/tree/master/examples/todo/src/repositories"><code>repositories/</code></a> directory</td>
+      <td>A sub-directory for all repository files. See <a href="Repository.html">Repository</a>.</td>
+      <td>&nbsp;</td>
+    </tr>
+    <tr>
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:83]: Should mention `src/__tests__`?

File: docs/site/Project-layout-reference.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+lang: en
+title: 'Project Layout Reference'
+keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Project-layout-reference.html
+---
+
+LoopBack 4 project files and directories are in the _application root
+directory_. Within this directory the standard LoopBack 4 project structure has
+these sub-directories:
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:11]: We should mention the project layout was created by `lb4 application` command and new artifacts can be added by other `lb4` commands or manually by application developers.

File: docs/site/Project-layout-reference.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
+---
+lang: en
+title: 'Project Layout Reference'
+keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Project-layout-reference.html
+---
+
+LoopBack 4 project files and directories are in the _application root
```
Comment (raymondfeng) [inline docs/site/Project-layout-reference.md:9]: We have two types of projects:

- Application project (for LB4 applications)
- Extension project (for LB4 extension modules)

Should we make it explicit this layout is for LB4 applications?

Comment (agnes512) [reply docs/site/Project-layout-reference.md:9]: Sure, I can change the title as well if we need another page for extension project.

File: docs/site/Loopback-application-layout-reference.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+LoopBack 4 application files and directories are in the _application root
+directory_. Within this directory the standard LB4 application structure has
+these sub-directories:
+
+- `src` - TypeScript source code and configuration files.
+- `public` - Client side assets (JavaScript, HTML, and CSS files) for the home
+  page.
+
+{% include tip.html content="By LoopBack naming conventions, artifacts such as models are grouped under their sub-directories. Also, with LoopBack artifact generators, they create class names in CamelCase and file names are in snake-case. For example, if you create a model named `MyUserModel` with the model generator, it creates files `my-user.model.ts` under `src/models`. See [Naming Convention](Command-line-interface.md#naming-convention) for more information.
```
Comment (dhmlau) [inline docs/site/Loopback-application-layout-reference.md:17]: Suggestion:
- `with LoopBack artifact generators, they create class names in CamelCase and file names are in snake-case.` -> `LoopBack artifact generators create class names in CamelCase and file names are in snake-case.`


```suggestion
{% include tip.html content="By LoopBack naming conventions, artifacts such as models are grouped under their sub-directories. Also, with LoopBack artifact generators, they create class names in CamelCase and file names are in snake-case. For example, if you create a model named `MyUserModel` with the model generator, it creates the file `my-user.model.ts` under `src/models` directory. See [Naming Convention](Command-line-interface.md#naming-convention) for more information.
```

File: docs/site/Loopback-application-layout-reference.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    </tr>
+    <tr>
+      <td><code>dist/</code> directory</td>
+      <td>Contains transpiled JavaScript code generated by <code>npm run build</code>.</td>
+      <td>npm run build</td>
+    </tr>
+    <tr>
+      <td><a href="https://github.com/strongloop/loopback-next/blob/master/examples/todo/package.json"><code>package.json</code></a></td>
+      <td>
+        Standard npm package specification. Use it to set up package dependencies, among other things. This file must be in the application root directory. See <a href="https://docs.npmjs.com/files/package.json"><code>npm-package.json</code> for information.</a>.
```
Comment (dhmlau) [inline docs/site/Loopback-application-layout-reference.md:44]: Duplicate `.` 

```suggestion
        Standard npm package specification. Use it to set up package dependencies, among other things. This file must be in the application root directory. See <a href="https://docs.npmjs.com/files/package.json"><code>npm-package.json</code> for information</a>.
```

File: docs/site/Loopback-application-layout-reference.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    </tr>
+    <tr>
+      <td><code>README.md</code></td>
+      <td>Stub file for internal documentation.</td>
+      <td>N/A</td>
+    </tr>
+    <tr>
+      <td><a href="https://github.com/strongloop/loopback-next/blob/master/examples/todo/tsconfig.json"><code>tsconfig.json</code></a></td>
+      <td>
+        A file specifies the root files and the compiler options required to compile the project. See <a href="https://www.typescriptlang.org/docs/handbook/tsconfig-json.html"><code>tsconfig.json</code> for information.</a>.
```
Comment (dhmlau) [inline docs/site/Loopback-application-layout-reference.md:56]: Duplicate `.` 

```suggestion
        A file specifies the root files and the compiler options required to compile the project. See <a href="https://www.typescriptlang.org/docs/handbook/tsconfig-json.html"><code>tsconfig.json</code> for information</a>.
```

File: docs/site/Project-layout-reference.md (hunk 13 above)
Comment (agnes512) [reply docs/site/Project-layout-reference.md]: 👍 I added CLI links to each artifact.

File: docs/site/Loopback-application-layout-reference.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
+---
+lang: en
+title: 'LoopBack 4 Application Layout Reference'
```
Comment (raymondfeng) [inline docs/site/Loopback-application-layout-reference.md:3]: Remove ` Reference`?

**Pull request merged**